reflex-dom-core/src/Reflex/Dom/Modals/Base.hs

```diff
@@ -0,0 +1,181 @@
+#ifndef ghcjs_HOST_OS
+instance MonadJSM m => MonadJSM (ModalsT t m)
+#endif
+
+withModalLayerClass :: forall t m a. (Reflex t, MonadFix m, DomBuilder t m, MonadHold t m)
+                    => Text
+                    -> ModalsT t m a
+                    -> m a
+withModalLayerClass c (ModalsT a) = elAttr "div" ("class" =: c) $ do
+  rec (result, requests) <- do
+        runRequesterT a modalDone
+      modalDone <- modalInner requests
+  return result
+
+modalInner :: forall t m k. (GCompare k, MonadHold t m, MonadFix m, DomBuilder t m)
+           => Event t (DMap k (Compose (ModalsT t m) (Event t)))
+           -> m (Event t (DMap k Maybe))
+modalInner rqs' = do
+  rec let rqs = fmap toInsertionsDMap rqs'
+          dones = fmap toDeletionsDMap modalDone
+      (m0, mpatch) <- traverseDMapWithKeyWithAdjust (\k v -> Compose <$> modalBody k v) DMap.empty (rqs <> dones)
+      modal <- fmap (merge . mapKeyValuePairsMonotonic wrapArgs) . incrementalToDynamic <$> holdIncremental m0 mpatch
+      let modalDone = fmap (mapKeyValuePairsMonotonic unwrapArgs) $ switch $ current modal
+  return modalDone
+ where
+  wrapArgs (k :=> v) = WrapArg k :=> getCompose v
+  unwrapArgs :: DSum (WrapArg Maybe k) Identity -> DSum k Maybe
+  unwrapArgs (WrapArg k :=> Identity v) = k :=> v
+  toInsertionsDMap :: DMap k v -> PatchDMap k v
+  toInsertionsDMap = coerce . DMap.map (ComposeMaybe . Just)
+  toDeletionsDMap :: forall k' v v'. DMap k' v -> PatchDMap k' v'
+  toDeletionsDMap = coerce . DMap.map (\(_ :: v a) -> ComposeMaybe (Nothing :: Maybe (v' a)))
+
+modalBody :: forall t m k a. (MonadHold t m, MonadFix m, DomBuilder t m)
+          => k a
+          -> Compose (ModalsT t m) (Event t) a
+          -> m (Event t (Maybe a))
+modalBody _ v = do
+  let style = T.intercalate ";"
+        [ "position:absolute"
+        , "left:0", "right:0", "top:0", "bottom:0"
+        , "background-color:rgba(0,0,0,0.1)"
+        , "display:flex", "justify-content:center", "align-items:center"
+        ]
+      backdropConfig = (def :: ElementConfig EventResult t (DomBuilderSpace m))
+        & elementConfig_eventSpec %~ addEventSpecFlags (Proxy :: Proxy (DomBuilderSpace m)) Click
+            (\_ -> stopPropagation)
+        & initialAttributes .~ ("style" =: style)
+  (overlay, complete) <- element "div" backdropConfig $ do
+    let bodyCfg = (def :: ElementConfig EventResult t (DomBuilderSpace m))
+          & elementConfig_eventSpec %~ addEventSpecFlags (Proxy :: Proxy (DomBuilderSpace m)) Click
+              (\_ -> stopPropagation)
+          & initialAttributes .~ ("style" =: "background-color:white;opacity:1;padding:1em")
+    fmap snd $ element "div" bodyCfg $ do
+      rec (rsps, rqs') <- runRequesterT (unModalsT (getCompose v)) innerDone
+          innerDone <- modalInner rqs'
+      return rsps
+  return $ leftmost
+    [ traceEventWith (const "butts") $ Just <$> complete
+    , traceEventWith (\_ -> "butts2") $ Nothing <$ domEvent Click overlay
```

Comment:
it's not clear to me either! after using it in anger i'm sure the most pressing needs will be obvious.